synapse/storage/databases/main/presence.py

```diff
@@ -54,19 +62,21 @@ def __init__(
         database: DatabasePool,
         db_conn: LoggingDatabaseConnection,
         hs: "HomeServer",
-    ):
+    ) -> None:
         super().__init__(database, db_conn, hs)
 
         self._can_persist_presence = (
             hs.get_instance_name() in hs.config.worker.writers.presence
         )
 
+        self._presence_id_gen: AbstractStreamIdGenerator
+
         if isinstance(database.engine, PostgresEngine):
             self._presence_id_gen = MultiWriterIdGenerator(
                 db_conn=db_conn,
                 db=database,
                 stream_name="presence_stream",
-                instance_name=self._instance_name,
+                instance_name=self._instance_name,  # type: ignore[attr-defined]
```

Comment:
I'd be sorely tempted to do a `self._instance_name = hs.get_instance_name()` at the top, even though it technically is redefines it

Reply:
Hopefully it's better now. The database query for a single user is not implemented `_get_presence_for_user`. My thoughts were that calling `get_presence_for_users` for every single user becomes a lot of overhead.
The loops are getting smaller now and perhaps increase the performance.